internal/graph/check.go

```diff
@@ -0,0 +1,485 @@
+
+	reducerFunc := cc.reduce(ctx, req, rewrite)
+
+	resolved, err := any(ctx, []ReduceableCheckFunc{reducerFunc})
+	resolved.Metadata = addCallToResponseMetadata(resolved.Metadata)
+
+	return resolved, err
+}
+
+// dispatch dispatches the provided request to the internal dispatcher and yields a value on
+// the result channel if the subproblem is successfully resolved or an error otherwise.
+func (cc *ConcurrentChecker) dispatch(req *dispatchpb.DispatchCheckRequest) ReduceableCheckFunc {
+	return func(ctx context.Context, resultChan chan<- *dispatchpb.DispatchCheckResponse) error {
+		result, err := cc.dispatcher.DispatchCheck(ctx, req)
+		resultChan <- result
+		return err
+	}
+}
+
+// reduce reduces the check request by breaking it into an appropriate subproblem that
+// is resolved by evaluating the underlying userset rewrite rule.
+func (cc *ConcurrentChecker) reduce(
+	ctx context.Context,
+	req *dispatchpb.DispatchCheckRequest,
+	rewrite *openfgapb.Userset,
+) ReduceableCheckFunc {
+	ctx, span := tracer.Start(ctx, "reduce")
+	defer span.End()
+
+	switch rw := rewrite.Userset.(type) {
+	case nil, *openfgapb.Userset_This:
+		span.SetAttributes(attribute.KeyValue{Key: "operation", Value: attribute.StringValue("this")})
+		return cc.checkDirect(ctx, req)
+	case *openfgapb.Userset_Union:
+		span.SetAttributes(attribute.KeyValue{Key: "operation", Value: attribute.StringValue("union")})
+		return cc.checkSetOperation(ctx, req, rw.Union.GetChild(), any)
+	case *openfgapb.Userset_Intersection:
+		span.SetAttributes(attribute.KeyValue{Key: "operation", Value: attribute.StringValue("intersection")})
+		return cc.checkSetOperation(ctx, req, rw.Intersection.GetChild(), all)
+	case *openfgapb.Userset_Difference:
+		span.SetAttributes(attribute.KeyValue{Key: "operation", Value: attribute.StringValue("difference")})
+		return func(ctx context.Context, resultChan chan<- *dispatchpb.DispatchCheckResponse) error {
+			// todo: wire me up properly
+			return fmt.Errorf("not implemented")
+		}
+	case *openfgapb.Userset_ComputedUserset:
+		span.SetAttributes(attribute.KeyValue{Key: "operation", Value: attribute.StringValue("computed")})
+		return cc.checkComputedUserset(ctx, req, rw)
+	case *openfgapb.Userset_TupleToUserset:
+		span.SetAttributes(attribute.KeyValue{Key: "operation", Value: attribute.StringValue("tuple-to-userset")})
+		return cc.checkTupleToUserset(ctx, req, rw)
+	default:
+		return func(ctx context.Context, resultChan chan<- *dispatchpb.DispatchCheckResponse) error {
+			return serverErrors.UnsupportedUserSet
+		}
+	}
+}
+
+func (cc *ConcurrentChecker) checkDirect(ctx context.Context, req *dispatchpb.DispatchCheckRequest) ReduceableCheckFunc {
+
+	return func(ctx context.Context, resultChan chan<- *dispatchpb.DispatchCheckResponse) error {
+		g := new(errgroup.Group)
+
+		g.Go(func() error {
+
+			tuple, err := cc.datastore.ReadUserTuple(ctx, req.WrappedRequest.StoreId, req.WrappedRequest.TupleKey)
```

Comment:
I think it's time for our storage interface to evolve and be refactored a bit. We have code all over the place that yields an `TupleIterator` only to throw away the tuple for the tuple key (e.g. `tuple.GetKey()`). So far I have seen 0 usage of the `tuple.Timestamp`, so I vote we refactor our storage interface in favor of a simpler and more elegant interface that just returns the things we actually need. This way we have less code that is an indirection to the thing that we actually want.